modules/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageClientsManager.java

```diff
@@ -0,0 +1,286 @@
+            }
+            final ProjectSecrets projectSecrets = project.custom(ProjectSecrets.TYPE);
+            // Project secrets can be null when node restarts. It may not have any s3 credentials if s3 is not in use.
+            if (projectSecrets == null || projectSecrets.getSettingNames().stream().noneMatch(key -> key.startsWith(GCS_SETTING_PREFIX))) {
+                // Most likely there won't be any existing client, but attempt to remove it anyway just in case
+                perProjectClientsHolders.remove(project.id());
+                continue;
+            }
+
+            final Settings currentSettings = Settings.builder()
+                // merge with static settings such as max retries etc
+                // TODO: https://elasticco.atlassian.net/browse/ES-11716 Consider change this to use per-project settings
+                .put(nodeGcsSettings)
+                .setSecureSettings(projectSecrets.getSettings())
+                .build();
+            final Map<String, GoogleCloudStorageClientSettings> clientSettings = GoogleCloudStorageClientSettings.load(currentSettings)
+                .entrySet()
+                .stream()
+                // Skip project clients that have no credentials configured. This should not happen in serverless.
+                // But it is safer to skip them and is also a more consistent behaviour with the cases when
+                // project secrets are not present.
+                .filter(entry -> entry.getValue().getCredential() != null)
+                .collect(Collectors.toUnmodifiableMap(Map.Entry::getKey, Map.Entry::getValue));
+
+            if (clientSettings.isEmpty()) {
+                // clientSettings should not be empty, i.e. there should be at least one client configured.
+                // But if it does somehow happen, log a warning and continue. The project will not have usable client but that is ok.
+                logger.warn("Skipping project [{}] with no client settings", project.id());
+                continue;
+            }
+
+            // TODO: If performance is an issue, we may consider comparing just the relevant project secrets for new or updated clients
+            // and avoid building the clientSettings
+            if (newOrUpdated(project.id(), clientSettings)) {
+                updatedPerProjectClients.put(project.id(), new PerProjectClientsHolder(clientSettings));
+            }
+        }
+
+        // Updated projects
+        for (var projectId : updatedPerProjectClients.keySet()) {
+            assert ProjectId.DEFAULT.equals(projectId) == false;
+            perProjectClientsHolders.put(projectId, updatedPerProjectClients.get(projectId));
+        }
+        // Removed projects
+        for (var projectId : perProjectClientsHolders.keySet()) {
+            if (currentProjects.containsKey(projectId) == false) {
+                assert ProjectId.DEFAULT.equals(projectId) == false;
+                perProjectClientsHolders.remove(projectId);
```

Comment:
The approach for managing per-project clients is similar to what we did for s3 but simpler. It is simpler because there is no ref-counting for the clients and we don't actively close it when it is not needed. Therefore, we simply remove and replace it when needed.